fix: If no routed members then do not fetch availability

joeauyeung opened the pull request:
## What does this PR do?

- When no routed team members are found, rather than falling back to all RR hosts, do not show availability

https://www.loom.com/share/0b37034efead4b87865c81cb824bcc17

Adds comprehensive unit tests for the routing form behavior change introduced in PR #25277, which prevents fallback to all members when attribute-based routing finds zero matching team members.

**Test Coverage Added:**
1. **findQualifiedHostsWithDelegationCredentials tests:**
   - Test for empty array with routing form response → returns no hosts (new behavior)
   - Test for null with routing form response → returns all hosts (existing behavior)
   - Test for empty array with contact owner → returns contact owner only
   
2. **getUrlSearchParamsToForward test:**
   - Test for empty `cal.routedTeamMemberIds` parameter when routing finds no matches
   - Test for `cal.routedTeamMemberIds=""` when `attributeRoutingConfig` is truthy but no matches found

3. **Bonus fix:**
   - Fixed pre-existing lint warning in test helper function (`hasOwnProperty` → `Object.prototype.hasOwnProperty.call()`)

## Updates since last revision

- Fixed test expectations in `getUrlSearchParamsToForward.test.ts` to match actual code behavior:
  - `cal.routedTeamMemberIds` is only added to URL params when `attributeRoutingConfig` is truthy OR when `teamMembersMatchingAttributeLogic` has values
  - When both `attributeRoutingConfig` and `teamMembersMatchingAttributeLogic` are null, the parameter is NOT included
  - Removed incorrect `"cal.routedTeamMemberIds": "null"` expectations from tests where `attributeRoutingConfig: null`

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a documentation change. **N/A** - test-only changes
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

Run the unit tests locally:

```bash
# Test the host filtering logic
TZ=UTC yarn test findQualifiedHostsWithDelegationCredentials.test.ts

# Test the URL parameter forwarding
TZ=UTC yarn test getUrlSearchParamsToForward.test.ts
```

**Expected Results:**
- All tests should pass
- New tests verify the null vs empty array distinction:
  - `routedTeamMemberIds: null` → shows all team members
  - `routedTeamMemberIds: []` → shows no availability (unless contact owner exists)

## Human Review Checklist

- [ ] Verify the distinction between `null` (no routing used) vs `[]` (routing used but found no matches) is correct for the use case
- [ ] Check that downstream consumers of `cal.routedTeamMemberIds` URL param handle the empty string value correctly when `attributeRoutingConfig` is truthy
- [ ] Verify the contact owner fallback behavior when `routedTeamMemberIds` is empty is correct
- [ ] Confirm that NOT including `cal.routedTeamMemberIds` when both `attributeRoutingConfig` and `teamMembersMatchingAttributeLogic` are null is the intended behavior


## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have checked if my changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works

---

**Link to Devin run**: https://app.devin.ai/sessions/7cc451a68eb2449d8fcab376ee9a51e5  
**Requested by**: joe@cal.com (@joeauyeung)

devin-ai-integration[bot] committed:
fix: prevent fallback to all members when routing finds no matches

When a routing form's attribute logic matches zero team members (empty array),
the booking page should show no availability instead of falling back to all members.

Changes:
- Modified findQualifiedHostsWithDelegationCredentials to return empty array when routedTeamMemberIds is explicitly empty
- Updated getUrlSearchParamsToForward to not add cal.routedTeamMemberIds parameter when array is empty
- Fixed pre-existing linting warning (any type) in getUrlSearchParamsToForward
- This ensures that when routing finds no matches, no members are loaded for the event

Co-Authored-By: joe@cal.com <j.auyeung419@gmail.com>

devin-ai-integration[bot] committed:
fix: handle null routedTeamMemberIds and prevent fallback when routing finds no matches

- Changed routedTeamMemberIds type to allow null to distinguish between 'not provided' and 'explicitly empty'
- Updated callers to pass null instead of converting to empty array
- Added early return guard when routing form explicitly returns empty array (no matches found)
- Added null check before calling .includes() on routedTeamMemberIds
- Reverted URL parameter change to keep cal.routedTeamMemberIds= even when empty (as requested)
- Fixed pre-existing linting warnings in util.ts (any -> unknown, removed unused catch variables)
- This ensures routing form with no matches shows no availability instead of falling back to all members

Co-Authored-By: joe@cal.com <j.auyeung419@gmail.com>

devin-ai-integration[bot] committed:
test: add tests for routed team members empty array behavior

- Add test for findQualifiedHostsWithDelegationCredentials when routingFormResponse is present and routedTeamMemberIds is empty array (should return no hosts)
- Add test for findQualifiedHostsWithDelegationCredentials when routingFormResponse is present but routedTeamMemberIds is null (should return all hosts)
- Add test for getUrlSearchParamsToForward when routing finds no matches (should forward cal.routedTeamMemberIds with empty value)
- Fix pre-existing lint warning in fromEntriesWithDuplicateKeys helper function (hasOwnProperty)

These tests verify the behavioral change where:
- null routedTeamMemberIds = no routing used → show all members
- empty array routedTeamMemberIds = routing found no matches → show no availability

Co-Authored-By: joe@cal.com <j.auyeung419@gmail.com>

joeauyeung commented on packages/features/bookings/lib/host-filtering/findQualifiedHostsWithDelegationCredentials.ts:
```diff
@@ -123,6 +123,10 @@ export class QualifiedHostsService {
     const fixedHosts = normalizedHosts.filter(isFixedHost);
     const roundRobinHosts = normalizedHosts.filter(isRoundRobinHost);
 
+    if (routingFormResponse && routedTeamMemberIds && routedTeamMemberIds.length === 0) {
```
Ensure that a `routingFormResponse` is also provided here

joeauyeung committed:
Merge branch 'main' into devin/no-fallback-empty-routed-members-1763569341

joeauyeung committed:
Pass `cal.routedTeamMemberIds=null` if no routed members found

joeauyeung committed:
Pass empty string instead of null

joeauyeung committed:
Still route to contact owner

joeauyeung committed:
Update tests

devin-ai-integration[bot] committed:
test: update test expectations to include cal.routedTeamMemberIds=null

Co-Authored-By: joe@cal.com <j.auyeung419@gmail.com>

joeauyeung committed:
Merge branch 'main' into devin/no-fallback-empty-routed-members-1763569341

joeauyeung commented on packages/features/routing-forms/lib/getUrlSearchParamsToForward.ts:
```diff
@@ -126,7 +126,9 @@ export function getUrlSearchParamsToForward({
       ? { ["cal.routedTeamMemberIds"]: teamMembersMatchingAttributeLogic.join(",") }
-      : null),
+      : attributeRoutingConfig
+        ? { ["cal.routedTeamMemberIds"]: "" }
```
If no routed team members are found then still return the param.

Udit-takkar reviewed (changes requested):
Unit tests are failing

devin-ai-integration[bot] committed:
fix: update test expectations to match new cal.routedTeamMemberIds behavior

Remove cal.routedTeamMemberIds: 'null' from test expectations where
attributeRoutingConfig is null, since the code only adds this parameter
when attributeRoutingConfig is truthy.

Co-Authored-By: joe@cal.com <j.auyeung419@gmail.com>
